File: tools/formatJson.py
import json, re

with open('assets/layouts/config.json', encoding='utf-8-sig') as f:
    data = json.load(f)

text = json.dumps(data, indent=2, ensure_ascii=False)

# Collapse 2-element numeric arrays onto one line
text = re.sub(r'\[\s*\n\s*(-?[\d.]+),\s*\n\s*(-?[\d.]+)\s*\n\s*\]', r'[\1, \2]', text)
# Collapse 3-element numeric arrays onto one line
text = re.sub(r'\[\s*\n\s*(-?[\d.]+),\s*\n\s*(-?[\d.]+),\s*\n\s*(-?[\d.]+)\s*\n\s*\]', r'[\1, \2, \3]', text)

# Collapse simple single-char string arrays (keyboard rows like ["Q","W","E",...])
def collapse_str_array(m):
    items = re.findall(r'"([A-Z0-9])"', m.group(0))
    return '[' + ', '.join(f'"{i}"' for i in items) + ']'

text = re.sub(r'\[(?:\s*\n\s*"[A-Z0-9]",?)+\s*\n\s*\]', collapse_str_array, text)

with open('assets/layouts/config.json', 'w', encoding='utf-8') as f:
    f.write(text)

print('Done. Lines:', text.count('\n'))

File: tools/formatJson.js
const fs = require('fs');
const path = require('path');

const filePath = path.resolve(__dirname, '../assets/layouts/config.json');
const json = JSON.parse(fs.readFileSync(filePath, 'utf8'));
let text = JSON.stringify(json, null, 2);

// Collapse 2-element numeric arrays onto one line (anchor, position, size, etc.)
text = text.replace(
  /\[\s*\n\s*(-?[\d.]+),\s*\n\s*(-?[\d.]+)\s*\n\s*\]/g,
  '[$1, $2]'
);

// Collapse 3-element numeric arrays onto one line
text = text.replace(
  /\[\s*\n\s*(-?[\d.]+),\s*\n\s*(-?[\d.]+),\s*\n\s*(-?[\d.]+)\s*\n\s*\]/g,
  '[$1, $2, $3]'
);

// Collapse single-letter/digit string arrays (keyboard rows of plain letters)
text = text.replace(
  /\[(\s*\n\s*"[A-Z0-9]",?)+\s*\n\s*\]/g,
  (m) => '[' + [...m.matchAll(/"([A-Z0-9])"/g)].map(x => `"${x[1]}"`).join(', ') + ']'
);

fs.writeFileSync(filePath, text, 'utf8');
console.log('Done. Lines:', text.split('\n').length);

